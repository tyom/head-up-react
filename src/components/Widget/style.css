.widget {
  position: absolute 0;
  margin: 4px;
  padding-top: calc(1.5rem + 1px);
  transition: 700ms;
  opacity: .7;
  border: 1px solid color(#fff a(.2));
  overflow: hidden;
}

.header {
  padding: 0 .5rem;
  line-height: 1.5rem;
  background-color: #1d1d1d;
  text-transform: uppercase;
  font-size: .8em;
  font-weight: 600;
  position: absolute 0 0 null 0;
  z-index: 1;
}

.body {
  overflow: hidden;
  position: relative;
  height: 100%;
}

.settings,
.body {
  transition: .2s;
  padding: .5rem;
  size: 100%;
}

.settings {
  background-color: #142a59;
  opacity: 0;
}

.menuBtn {
  background: none;
  border: 0;
  position: absolute;
  top: 3px;
  right: 0;
  color: #fff;
  padding: 0;
  outline: none;
  opacity: 0;
  transition: 200ms opacity;
  margin: 0 5px;

  &:hover .menuIcon {
    opacity: .7;
  }

  &:active .menuIcon {
    opacity: 1;
  }

  > span {
    display: hidden;
  }
}

.menuIcon {
  size: 1.1rem;
  opacity: .4;
}

.body svg {
  size: 100%;
}

/* Form */

fieldset {
  border: 0;
  padding: 0;
  margin: 0;
}

legend {
  margin: 0;
  padding: 0;
  display: block;
  font-weight: bold;
}

* + .field {
  margin-top: 10px;
}

.field label {
  display: table;
  opacity: .7;
}

.field input {
  background: color(#fff a(.5));
  border: 1px solid #000;
  padding: 4px;
  width: 100%;
  font-size: 1em;

  &:focus {
    outline: none;
    box-shadow: 0 0 0 1px #fff;
  }

  * + & {
    margin-top: 5px;
  }
}

/* States */

.is-configuring {
  .body {
    opacity: 0;
  }

  .settings {
    opacity: 1;
  }

  .settings,
  .body {
    transform: translateY(-100%);
  }
}

.is-active {
  border-color: color(#fff a(.5));
  box-shadow: 0 0 1px 1px color(#fff a(.1));
  opacity: 1;

  .menuBtn {
    opacity: 1;
  }
}
