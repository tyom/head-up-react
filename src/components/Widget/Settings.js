import React from 'react';
import PropTypes from 'prop-types';

import './style.css';

const defaultSettings = [];

function handleClick(evt, isActive) {
  if (isActive) {
    evt.stopPropagation();
  }
}

const Settings = ({ isActive, settings = [] }) => (
  <aside styleName="settings" onClick={evt => handleClick(evt, isActive)}>
    <fieldset>
      <legend>Settings</legend>
      {[...defaultSettings, ...settings].map(item => (
        <div styleName="field" key={item.label}>
          <label htmlFor={`${item.getWidgetId}-field-${item.name}`}>
            {item.label}
          </label>
          <input
            type="text"
            id={`${item.getWidgetId}-field-${item.name}`}
            value={item.value}
            onChange={item.onChange}
          />
        </div>
      ))}
    </fieldset>
  </aside>
);

Settings.propTypes = {
  isActive: PropTypes.bool,
  settings: PropTypes.array,
};

export default Settings;
