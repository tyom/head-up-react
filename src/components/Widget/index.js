import React, { Component } from 'react';
import PropTypes from 'prop-types';
import GoSettings from 'react-icons/lib/go/settings';
import classNames from 'classnames';
import Settings from './Settings';

import styles from './style.css';
import * as widgets from '../widgets';

const SettingsButton = ({ onClick }) => (
  <button styleName="menuBtn" onClick={onClick}>
    <GoSettings styleName="menuIcon" />
    <span>Menu</span>
  </button>
);

const Header = ({ title, onSettingsClick, showSettings }) => {
  function handleClick(evt) {
    const isButton = Boolean(evt.target.closest(`.${styles['menuBtn']}`));
    if (isButton) {
      evt.stopPropagation();
    }
    onSettingsClick && onSettingsClick();
  }

  return (
    <header styleName="header">
      {title}
      {showSettings ? <SettingsButton onClick={handleClick} /> : null}
    </header>
  );
};

const Body = ({ type, data }) => {
  const WidgetContent = widgets[type];

  return (
    <div styleName="body">
      {WidgetContent ? <WidgetContent {...data} /> : null}
    </div>
  );
};

class Widget extends Component {
  getSettingsFields(id) {
    return [
      {
        name: 'title',
        label: 'Title',
        value: this.state.title,
        getWidgetId: `w-${id}`,
        onChange: evt => {
          this.setState({
            title: evt.target.value,
          });
        },
      },
    ];
  }

  render() {
    const {
      id,
      type,
      data,
      title,
      isConfiguring,
      isActive,
      onClick,
      onSettingsClick,
    } = this.props;

    return (
      <article
        styleName={classNames('widget', {
          'is-configuring': isConfiguring,
          'is-active': isActive,
        })}
        onClick={onClick}
      >
        <Header
          id={id}
          title={title}
          onSettingsClick={() => onSettingsClick(id)}
          showSettings={isActive}
        />
        <Body type={type} data={data} />
        <Settings isActive={isActive} settings={this.getSettingsFields(id)} />
      </article>
    );
  }
}

SettingsButton.propTypes = {
  onClick: PropTypes.func,
};

Header.propTypes = {
  title: PropTypes.string,
  onSettingsClick: PropTypes.func,
  showSettings: PropTypes.bool,
};

Body.propTypes = {
  type: PropTypes.string,
  data: PropTypes.object,
};

Widget.propTypes = {
  id: PropTypes.string,
  title: PropTypes.string,
  type: PropTypes.string,
  data: PropTypes.object,
  onClick: PropTypes.func,
  isActive: PropTypes.bool,
  isConfiguring: PropTypes.bool,
  onSettingsClick: PropTypes.func,
  children: PropTypes.node,
};

export default Widget;
