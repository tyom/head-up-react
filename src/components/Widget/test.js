import React from 'react';
import Widget from './';

describe('Widget: structure', () => {
  test('initialise widget', () => {
    const wrapper = shallow(<Widget />);

    expect(wrapper).toMatchSnapshot();
  });
});

describe('Widget: behaviour', () => {
  test('click to activate cell', () => {
    const clickSpy = jest.fn();
    const wrapper = shallow(<Widget onClick={clickSpy} />);

    wrapper.simulate('click');

    expect(clickSpy).toHaveBeenCalled();
  });

  test('click to activate cell settings', () => {
    const clickSpy = jest.fn();
    const wrapper = mount(
      <Widget title="test" isActive={true} onSettingsClick={clickSpy} />
    );
    const settingsButton = wrapper.find('SettingsButton');

    settingsButton.simulate('click');

    expect(clickSpy).toHaveBeenCalled();
  });
});

describe('Cell: settings', () => {
  test('clicking settings calls a handler', () => {
    const clickSpy = jest.fn();
    const wrapper = shallow(<Widget onClick={clickSpy} />);

    wrapper.simulate('click');

    expect(clickSpy).toHaveBeenCalled();
  });
});