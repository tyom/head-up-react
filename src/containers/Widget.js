import { connect } from 'react-redux';

import { activateCell, activateCellSettings } from '../actions';
import Widget from '../components/Widget';

const mapStateToProps = (state, ownProps) => ({
  isActive: ownProps.id === state.headup.activeCell,
  isConfiguring: ownProps.id === state.headup.activeCellSettings,
});

export function mapDispatchToProps(dispatch, ownProps) {
  return {
    onClick: () => dispatch(activateCell(ownProps.id)),
    onSettingsClick: id => dispatch(activateCellSettings(id)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Widget);
