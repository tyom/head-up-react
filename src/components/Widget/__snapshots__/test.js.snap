// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget: structure initialise widget 1`] = `
<article
  className="style_widget__1AHS9"
>
  <Header
    onSettingsClick={[Function]}
  />
  <Body />
  <Settings
    settings={
      Array [
        Object {
          "getWidgetId": "w-undefined",
          "label": "Title",
          "name": "title",
          "onChange": [Function],
          "value": undefined,
        },
      ]
    }
  />
</article>
`;
